src/test/txrequest_tests.cpp

```diff
@@ -0,0 +1,539 @@
+                    if (prio_prev >= prio_cur) {
+                        ok = false;
+                        break;
+                    }
+                }
+                if (!ok) break;
+            }
+            if (ok) {
+                ok = m_runner.txhashset.insert(ret).second;
+            }
+        } while(!ok);
+        return ret;
+    }
+
+    GenTxid NewGTxid(const std::vector<std::vector<uint64_t>>& orders = {})
+    {
+        return {InsecureRandBool(), NewTxHash(orders)};
+    }
+
+    uint64_t NewPeer()
+    {
+        bool ok;
+        uint64_t ret;
+        do {
+            ret = InsecureRandBits(63);
+            ok = m_runner.peerset.insert(ret).second;
+        } while(!ok);
+        return ret;
+    }
+
+    std::chrono::microseconds Now() const { return m_now; }
+};
+
+/** Add to scenario a test with a single tx announced by a single peer.
+ *
+ * config is an integer between 0 and 31, which controls which variant of the test is used.
+ */
+void BuildSingleTest(Scenario& scenario, int config)
+{
+    auto peer = scenario.NewPeer();
+    auto gtxid = scenario.NewGTxid();
+    bool immediate = config & 1;
+    bool preferred = config & 2;
+    auto delay = immediate ? NO_TIME : RandomTime();
+
+    scenario.SetTestName(strprintf("Single(config=%i)", config));
+
+    // Receive an announcement, either immediately requestable or delayed.
+    scenario.ReceivedInv(peer, gtxid, preferred, immediate ? MIN_TIME : scenario.Now() + delay);
+    if (immediate) {
+        scenario.Check(peer, {gtxid}, 1, 0, 0, "s1");
+    } else {
+        scenario.Check(peer, {}, 1, 0, 0, "s2");
+        scenario.AdvanceTime(delay - MICROSECOND);
+        scenario.Check(peer, {}, 1, 0, 0, "s3");
+        scenario.AdvanceTime(MICROSECOND);
+        scenario.Check(peer, {gtxid}, 1, 0, 0, "s4");
+    }
+
+    if (config >> 3) { // We'll request the transaction
+        scenario.AdvanceTime(RandomTime());
+        auto expiry = RandomTime();
+        scenario.Check(peer, {gtxid}, 1, 0, 0, "s5");
+        scenario.RequestedTx(peer, gtxid, scenario.Now() + expiry);
+        scenario.Check(peer, {}, 0, 1, 0, "s6");
+
+        if ((config >> 3) == 1) { // The request will time out
+            scenario.AdvanceTime(expiry - MICROSECOND);
+            scenario.Check(peer, {}, 0, 1, 0, "s7");
+            scenario.AdvanceTime(MICROSECOND);
+            scenario.Check(peer, {}, 0, 0, 0, "s8");
+            return;
+        } else {
+            scenario.AdvanceTime(std::chrono::microseconds{InsecureRandRange(expiry.count())});
+            scenario.Check(peer, {}, 0, 1, 0, "s9");
+            if ((config >> 3) == 3) { // A reponse will arrive for the transaction
+                scenario.ReceivedResponse(peer, gtxid);
+                scenario.Check(peer, {}, 0, 0, 0, "s10");
+                return;
+            }
+        }
+    }
+
+    scenario.AdvanceTime(RandomTime());
```

Comment:
Isn't advancing the time blurring the further observance that the transaction state has been moved ? Compared to only relying on `DisconnectPeer`/`ForgetTxHash` ?

Reply:
I've made it conditional with an `if InsecureRandBool()` around it.